File: packages/sorapec/update.py
#!/usr/bin/env nix
#! nix shell --inputs-from .# nixpkgs#python3 --command python3

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "scripts"))

from updater import calculate_url_hash, fetch_github_latest_release, load_hashes, save_hashes, should_update

HASHES_FILE = Path(__file__).parent / "hashes.json"


def main() -> None:
    data = load_hashes(HASHES_FILE)
    current = data["version"]
    latest = fetch_github_latest_release("Kh05ifr4nD", "sorapec")

    print(f"Current: {current}, Latest: {latest}")

    if not should_update(current, latest):
        print("Already up to date")
        return

    url = f"https://github.com/Kh05ifr4nD/sorapec/releases/download/v{latest}/sorapec-{latest}-src.tar.gz"

    print("Calculating source hash...")
    source_hash = calculate_url_hash(url, unpack=True)

    save_hashes(HASHES_FILE, {"version": latest, "hash": source_hash})
    print(f"Updated to {latest}")


if __name__ == "__main__":
    main()

File: packages/sorapec/update.ts
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { assertString } from "../../scripts/updater/assert.ts";
import {
  calculateUrlHash,
  fetchGitHubLatestRelease,
  readJsonObjectFile,
  shouldUpdate,
  writeJsonFile,
} from "../../scripts/updater/module.ts";
import type { JsonValue } from "../../scripts/updater/module.ts";

async function main(): Promise<void> {
  const scriptDirectory = dirname(fileURLToPath(import.meta.url));
  const hashesFilePath = join(scriptDirectory, "hashes.json");

  const data = await readJsonObjectFile(hashesFilePath);
  const current = data["version"];
  assertString(current, `${hashesFilePath}: version must be a string`);

  const latest = await fetchGitHubLatestRelease("Kh05ifr4nD", "sorapec");

  console.log(`Current: ${current}, Latest: ${latest}`);

  if (!shouldUpdate(current, latest)) {
    console.log("Already up to date");
    return;
  }

  const url =
    `https://github.com/Kh05ifr4nD/sorapec/releases/download/v${latest}/sorapec-${latest}-src.tar.gz`;

  console.log("Calculating source hash...");
  const sourceHash = await calculateUrlHash(url, { unpack: true });

  const nextData: Record<string, JsonValue> = {
    version: latest,
    hash: sourceHash,
  };
  await writeJsonFile(hashesFilePath, nextData);

  console.log(`Updated to ${latest}`);
}

if (import.meta.main) {
  await main();
}
